app/components/data-source-menu.tsx

```diff
@@ -77,19 +80,45 @@ export const DataSourceProvider = ({ children }: { children: ReactNode }) => {
     setSource(source);
   };
 
+  const router = useRouter();
+
   useEffect(() => {
     const dataSource = retrieveDataSourceFromLocalStorage();
 
-    if (
-      dataSource &&
-      DATA_SOURCE_URLS.includes(stringifyDataSource(dataSource))
-    ) {
+    if (dataSource && SOURCES_BY_VALUE[stringifyDataSource(dataSource)]) {
       setSource(dataSource);
     } else {
       saveDataSourceToLocalStorage(DEFAULT_DATA_SOURCE);
     }
   }, []);
 
+  useEffect(() => {
+    if (router.isReady) {
+      const routerSourceLabel = router.query.dataSource as string;
+      const routerSource = SOURCES_BY_LABEL[routerSourceLabel]?.value;
+
+      const strSource = stringifyDataSource(source);
```

Comment:
```suggestion
      const providerSource = stringifyDataSource(source);
```

Reply:
I find the logic here a bit hard to follow. I think it would be beneficial to try to refactor this into a more meaningful chunk whose responsibility are well defined.

```
useSyncRouterParam({
  param: 'dataSource',
  value: providerSourceLabel,
  onParamChange: (routerParam) => handleSourceChange(parseDataSource(routerParam))
})
```

With the above example, we have clear boundaries between the thing knowing the router API and charged with keeping things in sync = useSyncRouterParam, and the thing that the provider knows, the business logic = the onParamChange function. What do you think ?